text/2978-stack_based_vec.md

```diff
@@ -0,0 +1,301 @@
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Instead of relying on a heap-allocator, stack-based memory area is added and removed on-demand in a last-in-first-out (LIFO) order according to the calling workflow of a program. Let's illustrate an imaginary function `super()` that allocates 32 bits before calling another function named `sub()` that also allocates 32 bits of auxiliary memory:
+
+```txt
+0 bits           32 bits          64 bits   
+
+
+|       |   ->   |       |   ->   |       |
+---------        ---------        ---------
+                 | super |        |  sub  |
+                 ---------        ---------
+                                  | super |
+                                  ---------
+```
+
+* Now `super()` returns and the exactly same thing happens in reverse order, dropping everything when out of scope.
+
+```txt
+64 bits          32 bits          0 bits   
+
+
+|       |   ->   |       |   ->   |       |
+---------        ---------        ---------
+|  sub  |        | super |
+---------        ---------
+| super |
+---------
+```
+
+`ArrayVec` takes advantage of this predictable behavior to reserve an exactly amount of uninitialized bytes up-front and these bytes form a buffer where elements can be included dynamically.
+
+```rust
+fn main() {
+    // `array_vec` has a pre-allocated memory of 2048 bits that can store up to 64 decimals.
+    let mut array_vec: ArrayVec<i32, 64> = ArrayVec::new();
+
+    // Although reserved, there isn't anything explicitly stored yet
+    assert_eq!(array_vec.len(), 0);
+
+    // Initializes the first 32 bits with a simple '1' decimal or
+    // 00000000 00000000 00000000 00000001 bits
+    array_vec.push(1);
+
+    // Our vector memory is now split into a 32/2016 pair of initialized and
+    // uninitialized memory respectively
+    assert_eq!(array_vec.len(), 1);
+}
+```
+
+Of course, fixed buffers lead to some inflexibility because unlike `Vec`, the underlying capacity can not expand at run-time and there will never be more than 64 elements in the above example.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`ArrayVec` is a contiguous memory block where elements can be collected, threfore, a collection by definition and even though `core::collections` doesn't exist, it is the most natural module placement.
+
+The API basically mimics most of the current `Vec` surface with some tweaks and additional methods to manage capacity. Notably, these additional methods are fallible versions of some well-known functions like `insert` that will return `Result` instead of panicking at run-time.
+
+```rust
+// core::collections
+
+pub struct ArrayVec<T, const N: usize> {
+    data: MaybeUninit<[T; N]>,
+    len: usize
+}
+
+impl<T, const N: usize> ArrayVec<T, N> {
+    // Constructors
+
+    pub unsafe fn from_raw_parts(_ptr: *mut T, _len: usize) -> Self;
```

Comment:
I don't feel that we should add this method here.

Reply:
Indeed